File: internal/scraper/locationtextures.go
package scraper

import (
	"context"
	"fmt"
	"log/slog"
	"regexp"
	"strings"

	"github.com/PuerkitoBio/goquery"
	"github.com/struffel/3d-assets-one/internal/model"
)

type LocationTexturesScraper struct{}

func (s *LocationTexturesScraper) Creator() model.Creator { return model.CreatorLocationTextures }

func (s *LocationTexturesScraper) ScrapeAssets(ctx context.Context, existing []model.StoredAsset) ([]ScrapedAsset, error) {
	const maxPerRun = 5
	const indexBase = "https://locationtextures.com/panoramas/free-panoramas/?page="

	tagRe := regexp.MustCompile(`[^A-Za-z0-9]`)
	var results []ScrapedAsset
	page := 1

	for {
		body, _, err := FetchURL(ctx, fmt.Sprintf("%s%d", indexBase, page))
		if err != nil {
			break
		}

		doc, err := goquery.NewDocumentFromReader(strings.NewReader(body))
		if err != nil {
			break
		}

		links := doc.Find("#product-category a.pack-link")
		if links.Length() == 0 {
			break
		}

		links.Each(func(_ int, sel *goquery.Selection) {
			if len(results) >= maxPerRun {
				return
			}

			href, _ := sel.Attr("href")
			imgEl := sel.Find("img.pack-link-img")
			imgTitle, _ := imgEl.Attr("title")
			imgSrc, _ := imgEl.Attr("data-src")

			if ContainsURL(existing, href) {
				return
			}

			// Fetch detail page for tags
			var detailTags []string
			detailBody, _, err := FetchURL(ctx, href)
			if err == nil {
				detailDoc, err := goquery.NewDocumentFromReader(strings.NewReader(detailBody))
				if err == nil {
					detailDoc.Find(`section a[href*="?tag"]`).Each(func(_ int, tagSel *goquery.Selection) {
						detailTags = append(detailTags, tagSel.Text())
					})
				}
			}

			// Tags from title + detail tags
			titleParts := tagRe.Split(imgTitle, -1)
			var tags []string
			for _, p := range titleParts {
				if p != "" {
					tags = append(tags, p)
				}
			}
			tags = append(tags, detailTags...)

			img, err := FetchImage(ctx, imgSrc)
			if err != nil {
				slog.Warn("locationtextures: thumbnail failed", "url", href, "err", err)
				return
			}

			results = append(results, ScrapedAsset{
				Title:        imgTitle,
				URL:          href,
				Tags:         tags,
				Type:         model.AssetTypeHDRI,
				Creator:      model.CreatorLocationTextures,
				Status:       model.ScrapedNewlyFound,
				RawThumbnail: img,
			})
		})

		if len(results) >= maxPerRun {
			break
		}
		page++
	}

	return results, nil
}

func (s *LocationTexturesScraper) ValidateAsset(ctx context.Context, asset model.StoredAsset) bool {
	return HeadOK(ctx, asset.URL)
}

File: app/creator/logic/location_textures.py
import re

from app.asset.asset_type import AssetType
from app.asset.scraped_asset import ScrapedAsset
from app.asset.scraped_asset_collection import ScrapedAssetCollection
from app.asset.scraped_asset_status import ScrapedAssetStatus
from app.asset.stored_asset_collection import StoredAssetCollection
from app.creator.creator import Creator
from app.creator.creator_logic import CreatorLogic
from app.fetch.web_item_reference import WebItemReference


class CreatorLogicLocationTextures(CreatorLogic):
    creator = Creator.LOCATION_TEXTURES
    max_assets_per_run = 5
    indexing_base_url = "https://locationtextures.com/panoramas/free-panoramas/?page="

    def scrape_assets(self, existing_assets: StoredAssetCollection) -> ScrapedAssetCollection:
        tmp_collection = ScrapedAssetCollection()
        page = 1
        processed = 0

        while True:
            dom = WebItemReference(self.indexing_base_url + str(page)).fetch().parse_as_dom()
            if dom is None:
                break

            pack_links = dom.select("#product-category a.pack-link")
            found_count = len(pack_links)

            for link in pack_links:
                href = link.get("href", "")
                img_tag = link.find("img", class_="pack-link-img")
                img_src = img_tag.get("data-src", "") if img_tag else ""
                title = img_tag.get("title", "") if img_tag else ""

                if existing_assets.contains_url(href):
                    continue

                # Fetch tags from the detail page
                detail_dom = WebItemReference(href).fetch().parse_as_dom()
                tags: list[str] = []
                if detail_dom:
                    tag_links = detail_dom.find_all("a", href=lambda h: h and "?tag" in h)
                    tags = [tl.get_text(strip=True) for tl in tag_links]

                title_tags = [t for t in re.split(r"[^A-Za-z0-9]", title) if t]
                asset = ScrapedAsset(
                    id=None,
                    creator_given_id=None,
                    title=title,
                    url=href,
                    tags=title_tags + tags,
                    type=AssetType.HDRI,
                    creator=Creator.LOCATION_TEXTURES,
                    status=ScrapedAssetStatus.NEWLY_FOUND,
                    raw_thumbnail=WebItemReference(url=img_src).fetch().parse_as_image() if img_src else None,
                )
                tmp_collection.append(asset)
                processed += 1

                if processed >= self.max_assets_per_run:
                    return tmp_collection

            page += 1
            if found_count == 0 or processed >= self.max_assets_per_run:
                break

        return tmp_collection
